src/components/Gallery.tsx

```diff
@@ -74,11 +76,12 @@ export default class Gallery extends Component<GalleryProps, GalleryState> {
     return urls;
   }
 
-  async componentDidUpdate(prevProps: GalleryProps) {
-    if (this.props.selectedSource !== prevProps.selectedSource) {
+  async componentDidUpdate() {
+    if (this.props.selectedSource.id !== this.state.currentSourceId) {
       const images = await this.getImagePaths();
       const fullUrls = this.constructUrl(images);
-      this.setState({ fullUrls });
+      const currentSourceId = this.props.selectedSource.id;
+      this.setState({ fullUrls, currentSourceId });
```

Comment:
Copying the selected source ID from props to state is an anti-pattern, see [the React docs](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-unconditionally-copying-props-to-state), I would strongly suggest reverting this change.

Reply:
So the docs say that it's unsafe to _unconditionally_ copy props to state, but here we are only doing it if there is a change that we are listening for, ie a different source is selected. Is it still considered an anti-pattern then?

Reply:
I would argue so, since there won't be any cases when `this.state.currentSourceId` won't be equal to `this.props.currentSourceId`.

Reply:
Hmm so maybe it's not clear, but the intention is that the two can be different:

1. User opens Dialog in empty state
2. User selects Source1
3. User re-selects Source1. At this point, `props.selectedSource == state.currentSource` so no request should be made for assets
3. User then selects Source2. At this point, `props.selectedSource != state.currentSource` so request is made for new assets

Although this may not be the most idiomatic approach, in which case I'm happy to rework it. Maybe it would be worth while to jump on a call to discuss a different approach.

Reply:
Firstly, this discussion is taking a bit too much effort for the benefit the resolution would bring, so feel free to drop this if you'd like.

---

That flow makes total sense to me. What I meant by the "won't be equal" comment is that after a full render cycle, `this.state. currentSourceId ` will always be equal to `this.props.currentSourceId`. In this function there is the case when they aren't the same, which is used to trigger the side-effect. That's totally fine. The thing is that this can be achieved without using state, and only using props, like so:

```jsx
  async componentDidUpdate(prevProps) {
    if (this.props.selectedSource.id !== prevProps.selectedSource.id) {
      const images = await this.getImagePaths();
      const fullUrls = this.constructUrl(images);
      this.setState({ fullUrls });
    }
  }
```

In general, the less state there is, the better

Reply:
Yep you're totally right. In fact I originally had the `prevProps` implementation but ran into some weird behavior, probably because the parent component was re-rendering too much. I'll go with your suggestion after I fix that underlying issue. Thanks for walking me through all that 👍